sdk/scheduler/src/main/java/com/mesosphere/sdk/framework/Driver.java

```diff
@@ -2,28 +2,34 @@
 
 import org.apache.mesos.SchedulerDriver;
 
-import java.util.Optional;
-
 /**
  * This class provides global access to a SchedulerDriver.
  */
 public final class Driver {
+
     private static SchedulerDriver driver;
-    private static Object lock = new Object();
 
     private Driver() {
         // Don't instantiate this class.
     }
 
-    public static Optional<SchedulerDriver> getDriver() {
-        synchronized (lock) {
-            return Optional.ofNullable(driver);
+    /**
+     * Returns the configured {@link SchedulerDriver} client object, or throws an {@link IllegalStateException} if no
+     * driver object is available. The returned Driver object should not be retained by the caller, as it may be
+     * replaced in the event of a re-registration.
+     */
+    public static SchedulerDriver getDriver() {
+        if (driver == null) {
+            throw new IllegalStateException("INTERNAL ERROR: No driver available");
         }
+        return driver;
     }
 
+    /**
+     * Assigns the driver object to be returned by calls to {@link #getDriver()}. In practice, this should be invoked
+     * when the scheduler first registers with Mesos, and/or following a re-registration.
+     */
     public static void setDriver(SchedulerDriver driver) {
-        synchronized (lock) {
-            Driver.driver = driver;
-        }
+        Driver.driver = driver;
```

Comment:
I ended up taking out the lock because of that. The lock wasn't really doing much and was giving a false sense of security if anything:
- Thread A calls getDriver() and gets copy of Driver pointer to execute query against
- Thread B calls setDriver(), driver pointer to new Driver
- Thread A is still using old Driver pointer...

Reply:
But now we have:
* Thread A calls `getInstance()` and uses it
* Thread B calls `setDriver()` and member is update to a new driver
* Thread A calls `getInstance()` and uses the new driver.

Does correctness for Thread A ever depend on both calls to `getInstance()` being the same instance -- e.g. when executed in a loop?

Reply:
Not that I'm aware of